File: sources/scala/xml/FactoryAdapter.java
package scala.xml;
// MIT License

import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.io.UnsupportedEncodingException;
import java.io.Writer;

import java.util.Map;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Stack;
import java.util.Iterator;

import org.xml.sax.Attributes;
import org.xml.sax.ContentHandler;


import org.xml.sax.ErrorHandler;
import org.xml.sax.Locator;

import org.xml.sax.SAXException;
import org.xml.sax.SAXNotRecognizedException;
import org.xml.sax.SAXNotSupportedException;
import org.xml.sax.SAXParseException;
import org.xml.sax.XMLReader;
import org.xml.sax.helpers.DefaultHandler;
import org.xml.sax.helpers.ParserAdapter;

import org.xml.sax.helpers.XMLReaderFactory;


/** 2do:
 * - handle MixedContent properly !
 *  - ? canonical XML ?
 *  - performance ( java Array <-> scala.Array )
 */
public abstract class FactoryAdapter
    extends DefaultHandler
    implements ContentHandler, ErrorHandler // SAX2
{

    // default settings

    /** Default parser name. */
    protected static final String
        DEFAULT_PARSER_NAME = "org.apache.xerces.parsers.SAXParser";

    /** Namespaces feature id (http://xml.org/sax/features/namespaces). */
    protected static final String
        NAMESPACES_FEATURE_ID = "http://xml.org/sax/features/namespaces";

    //
    // Constructors
    //

    /** Default constructor. */
    public FactoryAdapter() {
        buffer = new StringBuffer();
        attribStack = new Stack();
        hStack = new Stack();
        tagStack = new Stack();
        //fOut = new PrintWriter(System.out);
    } // <init>()


    // Data

    Stack hStack;   // [ element ] contains siblings
    Stack tagStack; // [String]

    Stack attribStack;

    String curTag;

    StringBuffer buffer;
    boolean capture;

    // these are defined in a scala class

    /** Tests if an XML element contains text.
     * @return true if element named <code>localName</code> contains text.
     */
    abstract boolean elementContainsText(String localName);

    /** Creates an new XML element.
     * @param elemName
     * @param attribs
     * @param chIter
     * @return a new XML element.
     */
    abstract scala.Object createElement(String elemName,
                                        Map attribs,
                                        Iterator chIter);

    /** Creates an PCDATA element.
     * @param text
     * @return a new PCDATA element.
     */
    abstract scala.Object createPCDATA(String text);

    //
    // ContentHandler methods
    //

    /** Set document locator.
     * @param locator
     */
    public void setDocumentLocator(Locator locator) {
    }

    /** Start document.
     * @throws org.xml.sax.SAXException if ..
     */
    public void startDocument() throws SAXException {
    }


    /** Characters.
     * @param ch
     * @param offset
     * @param length
     */
    public void characters(char[] ch, int offset, int length)
            throws SAXException {

        boolean ws = false;

        //System.err.println( "before \""+buffer+"\"" );

        if (capture)
            for (int i = offset; i < offset + length; i++)
                if (Character.isWhitespace(ch[i])) {
                    if (!ws) {
                        buffer.append(' ');
                        ws = true;
                    }
                } else {
                    buffer.append(ch[i]);
                    ws = false;
                }

        //System.err.println( "after \""+buffer+"\"" );
        //System.err.println();

    } // characters(char[],int,int)

    /** Ignorable whitespace.
     */
    public void ignorableWhitespace(char[] ch, int offset, int length)
        throws SAXException {
    }

    /** End document.
     */
    public void endDocument() throws SAXException {
    }

    //
    // ContentHandler methods
    //

    /** Start prefix mapping.
     */
    public void startPrefixMapping(String prefix, String uri)
        throws SAXException {
    }

    /** Start element.
     */
    public void startElement(String uri, String localName, String qname,
                             Attributes attributes) throws SAXException {

        captureText();

        tagStack.push(curTag);
        curTag = localName ;

        capture = elementContainsText(localName) ;

        hStack.push( null );
        HashMap map = null;

        if (attributes == null) {
              //fOut.println("null");
        }
        else {
              map = new HashMap();

              int length = attributes.getLength();
              for (int i = 0; i < length; i++) {
                    String attrLocalName = attributes.getLocalName(i);
                    //String attrQName = attributes.getQName(i);
                    //String attrURI = attributes.getURI(i);
                    String attrType = attributes.getType(i);
                    String attrValue = attributes.getValue(i);
                    if( attrType.equals("CDATA") )
                          map.put( attrLocalName, attrValue );

              }
        }
        attribStack.push( map );

    } // startElement(String,String,String,Attributes)


    /** this way to deal with whitespace is quite nonstandard, but useful
     */
    void captureText() {
        if (capture == true) {
             hStack.push(createPCDATA(buffer.toString()));
        }
        buffer.setLength(0);
    }

    /** End element.
     * @param uri
     * @param localName
     * @param qname
     * @throws org.xml.sax.SAXException if ..
     */
    public void endElement(String uri, String localName, String qname)
            throws SAXException {

        captureText();

        Map attribMap = (HashMap) attribStack.pop();

        // reverse order to get it right
        LinkedList v       = new LinkedList();
        scala.Object child = (scala.Object) hStack.pop();
        while( child != null ) {
            v.addFirst(child);
            child = (scala.Object) hStack.pop();
        }

        // create element
        rootElem = createElement(localName, attribMap, v.iterator());
        hStack.push(rootElem);

        // set
        curTag = (String) tagStack.pop();

        if (curTag != null) // root level
            capture = elementContainsText(curTag);
        else
            capture = false;

    } // endElement(String,String,String)

    /** End prefix mapping.
     */
    public void endPrefixMapping(String prefix) throws SAXException {
    }

    /** Skipped entity.
     */
    public void skippedEntity(String name) throws SAXException {
    }

    //
    // ErrorHandler methods
    //

    /** Warning.
     */
    public void warning(SAXParseException ex) throws SAXException {
        printError("Warning", ex);
    }

    /** Error.
     */
    public void error(SAXParseException ex) throws SAXException {
        printError("Error", ex);
    }

    /** Fatal error.
     */
    public void fatalError(SAXParseException ex) throws SAXException {
        printError("Fatal Error", ex);
        throw ex;
    }

    //
    // Protected methods
    //

    /** Prints the error message.
     */
    protected void printError(String type, SAXParseException ex) {

        System.err.print("[");
        System.err.print(type);
        System.err.print("] ");

        String systemId = ex.getSystemId();
        if (systemId != null) {
            int index = systemId.lastIndexOf('/');
            if (index != -1)
                systemId = systemId.substring(index + 1);
            //System.err.print(systemId);
        }

        System.err.print(':');
        System.err.print(ex.getLineNumber());
        System.err.print(':');
        System.err.print(ex.getColumnNumber());
        System.err.print(": ");
        System.err.print(ex.getMessage());
        System.err.println();
        System.err.flush();

    } // printError(String,SAXParseException)

    scala.Object rootElem;

    //FactoryAdapter
    // MAIN
    //

    /** load XML document
     * @param fileName
     * @return a new XML document object
     */
    public scala.Object loadXML(String fileName) {

        // variables
        //PrintWriter out = new PrintWriter(System.out);
        XMLReader parser = null;

        // use default parser
        // create parser
        try {
             parser = XMLReaderFactory.createXMLReader(DEFAULT_PARSER_NAME);
        }
        catch (Exception e) {
            System.err.println("error: Unable to instantiate parser (" +
                               DEFAULT_PARSER_NAME + ")");
            System.exit(-1);
        }

        // set parser features
        try {
            parser.setFeature(NAMESPACES_FEATURE_ID, true);
        }
        catch (SAXException e) {
            System.err.println("warning: Parser does not support feature (" +
                               NAMESPACES_FEATURE_ID + ")");
        }

        // set handlers
        parser.setErrorHandler(this);

        if (parser instanceof XMLReader) {
            parser.setContentHandler(this);
        }
        else {
            System.err.println("wrong parser");
            System.exit(-1);
        }

        // parse file
        try {
            System.err.println("[parsing \"" + fileName + "\"]");
            parser.parse( fileName );
        }
        catch (SAXParseException e) {
            // ignore
        }
        catch (Exception e) {
            System.err.println("error: Parse error occurred - " + e.getMessage());
            if (e instanceof SAXException) {
                e = ((SAXException)e).getException();
            }
            e.printStackTrace(System.err);
        }

        return rootElem;

    } // loadXML

} // class FactoryAdapter

File: src/library/scala/xml/parsing/FactoryAdapter.scala
/*                     __                                               *\
**     ________ ___   / /  ___     Scala API                            **
**    / __/ __// _ | / /  / _ |    (c) 2003-2011, LAMP/EPFL             **
**  __\ \/ /__/ __ |/ /__/ __ |    http://scala-lang.org/               **
** /____/\___/_/ |_/____/_/ | |                                         **
**                          |/                                          **
\*                                                                      */

package scala.xml
package parsing

import java.io.{ InputStream, Reader, File, FileDescriptor, FileInputStream }
import scala.collection.{ mutable, Iterator }
import org.xml.sax.Attributes
import org.xml.sax.helpers.DefaultHandler

// can be mixed into FactoryAdapter if desired
trait ConsoleErrorHandler extends DefaultHandler {
  // ignore warning, crimson warns even for entity resolution!
  override def warning(ex: SAXParseException): Unit = { }
  override def error(ex: SAXParseException): Unit = printError("Error", ex)
  override def fatalError(ex: SAXParseException): Unit = printError("Fatal Error", ex)

  protected def printError(errtype: String, ex: SAXParseException): Unit =
    Console.withOut(Console.err) {
      val s = "[%s]:%d:%d: %s".format(
        errtype, ex.getLineNumber, ex.getColumnNumber, ex.getMessage)
      Console.println(s)
      Console.flush
    }
}

/** SAX adapter class, for use with Java SAX parser. Keeps track of
 *  namespace bindings, without relying on namespace handling of the
 *  underlying SAX parser.
 */
abstract class FactoryAdapter extends DefaultHandler with factory.XMLLoader[Node] {
  var rootElem: Node = null

  val buffer      = new StringBuilder()
  val attribStack = new mutable.Stack[MetaData]
  val hStack      = new mutable.Stack[Node]   // [ element ] contains siblings
  val tagStack    = new mutable.Stack[String]
  var scopeStack  = new mutable.Stack[NamespaceBinding]

  var curTag : String = null
  var capture: Boolean = false

  // abstract methods

  /** Tests if an XML element contains text.
   * @return true if element named `localName` contains text.
   */
  def nodeContainsText(localName: String): Boolean // abstract

  /** creates an new non-text(tree) node.
   * @param elemName
   * @param attribs
   * @param chIter
   * @return a new XML element.
   */
  def createNode(pre: String, elemName: String, attribs: MetaData,
                 scope: NamespaceBinding, chIter: List[Node]): Node // abstract

  /** creates a Text node.
   * @param text
   * @return a new Text node.
   */
  def createText(text: String): Text // abstract

  /** creates a new processing instruction node.
  */
  def createProcInstr(target: String, data: String): Seq[ProcInstr]

  //
  // ContentHandler methods
  //

  val normalizeWhitespace = false

  /** Characters.
  * @param ch
  * @param offset
  * @param length
  */
  override def characters(ch: Array[Char], offset: Int, length: Int): Unit = {
    if (!capture) return
    // compliant: report every character
    else if (!normalizeWhitespace) buffer.appendAll(ch, offset, length)
    // normalizing whitespace is not compliant, but useful
    else {
      var it = ch.slice(offset, offset + length).iterator
      while (it.hasNext) {
        val c = it.next
        val isSpace = c.isWhitespace
        buffer append (if (isSpace) ' ' else c)
        if (isSpace)
          it = it dropWhile (_.isWhitespace)
      }
    }
  }

  private def splitName(s: String) = {
    val idx = s indexOf ':'
    if (idx < 0) (null, s)
    else (s take idx, s drop (idx + 1))
  }

  /* ContentHandler methods */

  /* Start element. */
  override def startElement(
    uri: String,
    _localName: String,
    qname: String,
    attributes: Attributes): Unit =
  {
    captureText()
    tagStack push curTag
    curTag = qname

    val localName = splitName(qname)._2
    capture = nodeContainsText(localName)

    hStack push null
    var m: MetaData = Null
    var scpe: NamespaceBinding =
      if (scopeStack.isEmpty) TopScope
      else scopeStack.top

    for (i <- 0 until attributes.getLength()) {
      val qname = attributes getQName i
      val value = attributes getValue i
      val (pre, key) = splitName(qname)
      def nullIfEmpty(s: String) = if (s == "") null else s

      if (pre == "xmlns" || (pre == null && qname == "xmlns")) {
        val arg = if (pre == null) null else key
        scpe = new NamespaceBinding(arg, nullIfEmpty(value), scpe)
      }
      else
        m = Attribute(Option(pre), key, Text(value), m)
    }

    scopeStack push scpe
    attribStack push m
  }


  /** captures text, possibly normalizing whitespace
   */
  def captureText(): Unit = {
    if (capture && buffer.length > 0)
      hStack push createText(buffer.toString)

    buffer.clear()
  }

  /** End element.
   * @param uri
   * @param _localName
   * @param qname
   * @throws org.xml.sax.SAXException if ..
   */
  override def endElement(uri: String , _localName: String, qname: String): Unit = {
    captureText()
    val metaData = attribStack.pop

    // reverse order to get it right
    val v = (Iterator continually hStack.pop takeWhile (_ != null)).toList.reverse
    val (pre, localName) = splitName(qname)
    val scp = scopeStack.pop

    // create element
    rootElem = createNode(pre, localName, metaData, scp, v)
    hStack push rootElem
    curTag = tagStack.pop
    capture = curTag != null && nodeContainsText(curTag) // root level
  }

  /** Processing instruction.
  */
  override def processingInstruction(target: String, data: String) {
    hStack pushAll createProcInstr(target, data)
  }
}
